Adding an example on video classification

sayakpaul opened the pull request:
There are many subtleties for training a well-performing video classifier. Also, there are many ways to train one. This example walks through one of them. Hopefully, it will be helpful for the community.

fchollet commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,434 @@
+"""
+Title: Video Classification with a Conv-LSTM Architecture
```
CNN-RNN

fchollet reviewed (commented):
Thanks for the PR. This is great stuff! Do you have a notebook version so we can take a look at the visualizations?

sayakpaul commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,434 @@
+In this example, we will be using such a hybrid architecture consisting of a
+Convolutional Neural Network (CNN) and a Recurrent Neural Network (RNN) consisting of
+[GRU layers](https://keras.io/api/layers/recurrent_layers/gru/). These kinds of hybrid
+architectures are popularly known as **Conv-LSTM**.
```
Oh! Did not know about this lingo. Thank you!

sayakpaul commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,434 @@
+# First we represent the class labels as integers and then we
+# one-hot encode them.
+train_labels = lb.fit_transform(new_train_df["class"].values.tolist())
+test_labels = lb.transform(new_test_df["class"].values.tolist())
```
Thanks for the suggestions. Settled with `StringLookup`.

sayakpaul commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,434 @@
+
+
+def build_feature_extractor():
+    feature_extractor = keras.applications.InceptionV3(
```
Yes. Added this to the notes since I wanted to follow some recipes from the existing works. Label smoothing, using InceptionV3 are some of them.

sayakpaul commented:
Thank you, @fchollet for the review. I have addressed all your comments. Here's a [Colab Notebook](https://colab.research.google.com/drive/1PUIgFTWFcCpJdIeUWKYLCGQOODs_PnGb?usp=sharing) for visualizations.

fchollet commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,466 @@
+    rnn_model = keras.Sequential(
+        [
+            keras.layers.GRU(
+                64, input_shape=(SEQ_LENGTH, 5 * 5 * 2048), return_sequences=True
```
If you really want to provide an input shape, rather than using the `input_shape` argument prefer using an `Input()` object as first layer.

fchollet reviewed (commented):
Thanks for the update!

sayakpaul commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,469 @@
+"""
+
+label_processor = keras.layers.experimental.preprocessing.StringLookup(
+    max_tokens=6, num_oov_indices=0, output_mode="binary", sparse=False
```
Referring to [this guide](https://www.tensorflow.org/tutorials/structured_data/preprocessing_layers#categorical_columns) it seems like we need to first run the labels `StringLookup` and then use `CategoryEncoding` since `CategoryEncoding` cannot directly operate strings. This is why I directly used `StringLookup`.

fchollet replied:
I was confused, you're right. It looks good.

Since you're calling `adapt` to find out the label set, then you don't need to specify `max_tokens` (also `sparse=False` is the default so no need to specify it).

Another thing you could do is skip calling `adapt` altogether, by doing:

```python
label_processor = keras.layers.experimental.preprocessing.StringLookup(
    num_oov_indices=0, output_mode="binary",
    vocabulary=new_train_df["class"].values
)
```

That way the layer will already have its vocab ready and won't need `adapt` anymore.

sayakpaul replied:
Great. Will apply this right away.

sayakpaul replied:
@fchollet, after trying the suggestion, I got this:

```
ValueError: The passed vocabulary has at least one repeated term. Please uniquify your dataset. The repeated terms are ['CricketShot', 'PlayingCello', 'Punch', 'ShavingBeard', 'TennisSwing']
```

sayakpaul commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,466 @@
+broken_video_names_train = separate_frames(train_video_names, "train", "train_frames")
+broken_video_names_test = separate_frames(test_video_names, "test", "test_frames")
+
+"""
```
This sounds great. But the first step is not clear to me. I am suspecting the changes need to take place in `separate_frames()`.  If you could provide a short example that would be helpful.

sayakpaul replied:
I was able to figure out the network structure, though:

```python
def build_feature_extractor():
    feature_extractor = keras.applications.InceptionV3(weights="imagenet", include_top=False,
                                                       input_shape=(IMG_SIZE, IMG_SIZE, 3))
    feature_extractor.trainable = False

    inputs = keras.Input((SEQ_LENGTH, IMG_SIZE, IMG_SIZE, 3))
    # preprocessed = keras.applications.inception_v3.preprocess_input(inputs)

    outputs = keras.layers.TimeDistributed(feature_extractor)(inputs)
    outputs_flat = keras.layers.Reshape((SEQ_LENGTH, 5*5*2048))(outputs)
    
    return keras.Model(inputs, outputs_flat, name="feature_extractor")
```

But notice that I couldn't figure a way out to apply `preprocess_input()` to the inputs now. Looking forward to your suggestions.

fchollet replied:
Excellent, yes this is the expected architecture.

> But notice that I couldn't figure a way out to apply preprocess_input() to the inputs now

In the current code, you're building a Functional model that contains `preprocess_input`. So you could just replace it with `TimeDistributed (Lambda(preprocess_input))`. Another option would be to `map` into the dataset a function that reshapes the data to `(-1, 224, 224, 3)` then apply `preprocess_input` then reshape back to the original shape.

sayakpaul replied:
Let me look into these. I will reach back out if I am stuck. Thanks, @fchollet.

fchollet commented on examples/vision/video_classification.py:
```diff
@@ -0,0 +1,466 @@
+those integer values.
+"""
+
+nb_classes = len(np.unique(new_train_df["class"].values))
```
Use `num_` as numbering prefix for consistency with other examples and with the API

sayakpaul commented:
@fchollet, I tried to incorporate [your suggestions](https://github.com/keras-team/keras-io/pull/478#discussion_r642120290) but it weirdly affects the performance quite a bit. I am unsure as to why this might be the case. Would you be able to take a look at [the notebook](https://colab.research.google.com/gist/sayakpaul/5c8e9438736ec9a765588883cc4222b1/video_classification.ipynb)?

fchollet reviewed (commented):
I trying running the notebook with a few simplifications:

- apply pooling in the feature extractor to limit the number of features
- simplify the classification model
- shuffle the training data and use a validation split to monitor overfitting

Notebook: https://colab.research.google.com/drive/1sfkoKxEF_kfGU1vi-_rNOQeu_Ydx1EtF?usp=sharing

The results I'm seeing are consistent with what I'd expect given the small number of samples and the large model sizes: quick overfitting and low generalization.

To train a model with this many parameters (the classification model you had has 10M parameters) you'd need tens of thousands of samples at the very least. Right now there are 152 training samples. This is just impossible. I recommend both simplifying the model (e.g. some of the changes outlined above) and increasing the size of the training data by at least 10x.

If you get to 1000 training samples you can probably manage to train a classification model with 10-100k parameters.

sayakpaul commented:
Thank you for your input, @fchollet. I realized I had introduced a pesky bug during preparing the frame sequences. This was hurting the mappings between the videos and labels. I took care of it and incorporated the following suggestions:

* Used pooling inside the `feature_extractor`.
* Simplified the sequence models (now has 99099 parameters). 
* Made an extensive note on the use of a small dataset with respect to using a relatively larger model. 

With these changes, the performance is good enough. Here's [the notebook](https://colab.research.google.com/github/sayakpaul/Action-Recognition-in-TensorFlow/blob/main/Video_Classification.ipynb). 

Additionally, your contributions to this would-be example are non-trivial IMHO. They go way beyond just code reviews and improvement suggestions. Therefore, I would very much welcome the idea of adding you as a co-author for the example. LMK.

fchollet commented:
Glad you were able to fix the bug!

I've been thinking about this example some more. This would be our first video classification example, which is great. But that means it should also exemplify best practices -- techniques that will work well across most video classification problems, not just this specific dataset. Right now, some things are very good, like the general idea of using a CNN for feature extraction and then a RNN. However, some things could be improved. Here are some things I'd like to see in a generic video classification example:

1. Use a fixed frame subsampling rate, and have videos of different length.

There's no real justification for using exactly N frames per video. In the real world your videos will have different lengths. The justification "we need to batch images" doesn't work because you can just pad shorter videos with zeros and generate a padding mask, which the RNN will take into account (it will skip masked frames, so it won't even slow down training).

How that would work in practice: use a batch size of 1 during feature extraction, then pad the vector sequences with zeros and add a mask input to the classification model, which would go in the `mask` argument in the first RNN call.

2. Use a higher frame count.

5 frames is very low and doesn't justify the use of a RNN as the importance of order will be extremely marginal. We need the inputs to be actual videos, not a small collection of pictures, so more like 20+ frames.

3. Use more training samples.

As mentioned, at least a few hundreds per class. If you have 5 balanced classes then 1000-2000 is a reasonable number.

> Therefore, I would very much welcome the idea of adding you as a co-author for the example.

Don't worry about it, I'm just the janitor here.

Let me know.

sayakpaul commented:
Thanks!

1> Could you provide a minimal code example / relevant resource for this?

2> Alright. 

3> Alright. But also note that because of broken videos (videos that OpenCV could not capture) the number of available videos got reduced further.

fchollet commented:
> Could you provide a minimal code example

I imagine you could do something like (untested code, just writing it up here):

```python
frame_masks = np.zeros(shape=(num_samples, max_seq_length), dtype='uint8')
frame_features = np.zeros(shape=(num_samples, max_seq_length, num_features), dtype='float32')
for i, batch in enumerate(dataset):  # Dataset is batched with size 1
   length = batch.shape[1]  # This is different from video to video
   for j in range(length):
     frame_features[i, j, :] = feature_extractor(batch[:, j, :])  # feature_extractor is just a CNN that returns vectors of features
   frame_masks[i, :length] = 1  # 1 = not masked, 0 = masked

... # later

frame_features_input = Input((max_seq_length, num_features))
mask_input = Input((max_seq_length,), dtype='uint8')
x = GRU(...)(frame_features_input, mask=mask_input)

... # later
model = Model([frame_features_input, mask_input], output)
model.fit([frame_features, frame_masks], labels, ...)
```

Does that make sense? Can't guarantee it will work exactly as is, but it gives you an idea of what is supposed to happen.

sayakpaul commented:
Thanks for providing the snippet. What is the shape of a single entry inside `dataset`?

fchollet commented:
> It seems to work but it's very slow

I think this might be because you won't get multiprocessing on Colab. But regardless of the environment I strongly recommend using a `Dataset` to avoid such performance issues (or just NumPy arrays, which amounts to the same). It will also enable you to train on TPU, etc. All the data is available as arrays so it seems doable.

A random thing you can try to diagnose the issue is to use `from_generator` to make a `Dataset` and then call `cache()` on it. See if it improves performance for the next epoch.

> is it possible to limit the number of examples to the current setting in the interest of the runtime

Is there an obstacle with getting more data, or the issue the run time?

sayakpaul commented:
> I think this might be because you won't get multiprocessing on Colab.

I tried it on a GCP VM (N1-standard-8) too but the performance was roughly the same. 

> or just NumPy arrays, which amounts to the same

I am more inclined toward creating a standard dataset for this since we are trying to showcase a couple of best practices here. Besides, showing the readers how to create a standard video classification dataset as a part of the example would be beneficial too. 

> Is there an obstacle with getting more data, or the issue the run time?

Runtime is currently the main issue. 

Additionally, if you could check the notebook in my previous comment and let me know if it's close to what you had envisioned that would be helpful.

fchollet commented:
I checked out the notebook and the model looks great!

> I am more inclined toward creating a standard dataset for this

Do you mean an end-to-end tf.data dataset where you provide video filenames and labels and it gives you an iterable with encoded video frames, masks, and encoded labels? This is certainly doable. To avoid doing redundant feature extraction work you're going to want to use caching though. I would encourage you to go this route (in particular to preserve TPU compatibility).

Actually this is currently a major issue with your `Sequence` class that may explain most of the slowdown: you apply feature extraction for every batch, while training proceeds. This means you're doing a ton of extra work in later epochs. In the previous workflow, you did feature extraction once, beforehand. This represents a large difference in GPU compute load.

fchollet commented:
> Runtime is currently the main issue.

Then let's increase the data size and let's reduce the number of epochs (while documenting performance achieved for the full number of epochs).

sayakpaul commented:
Okay let me look into these. 

TPU compatibility is questionable if we want to extract the frames using OpenCV, though. I would do that using tf.py_funnction() which is unsupported with TPUs as far as I remember.

fchollet commented:
> TPU compatibility is questionable if we want to extract the frames using OpenCV, though

You could do this, but I would recommend doing the Python-only parts of the preprocessing beforehand, then using `from_tensor_slices` to create a dataset (if the data is too large you can save processed files to disk first). The CNN feature extraction part can either be done beforehand or as part of the dataset, though beforehand is actually better because otherwise you need to remember to call `.cache()` if you don't want to do dramatically more work (kind of a trap).

sayakpaul commented:
> Does it make sense it this case to attempt to parallelize the preprocessing? If most of the time spent is OpenCV, and if that algo is already parallel, then there's not much we can do.

I was going to suggest this but the OpenCV backend utilities are already threaded. So, there's nothing much we can do here.

> Also, it seems we can increase the epoch number if training is so fast.

yes, totally.

> I think if we keep the total runtime below 30min it will be a success. How many training samples would that get us?

A few maybe 30 or 50. But that risks the runtime being overloaded with the preprocessing we are doing.

sayakpaul commented:
@fchollet let me know how you'd want us to proceed from here.

fchollet commented:
@sayakpaul I'll let you make the call with regard to how many samples to use -- more data is better, but we should try to keep the runtime below 30 min. Up to you at this point!

A random thought: by using a small/faster CNN you'll be able to cut *some* processing time (but not too much if it's mostly OpenCV time).

sayakpaul commented:
Thanks, @fchollet. Let me incorporate these points in the next commit. This PR has been an amazing learning experience for me. Thank you for pushing it forward :)

sayakpaul commented:
@fchollet added changes to the example. Here is the summary:

* Currently, the example operates with the videos that belong to the top-5 classes present in the UCF101 dataset. Increasing the top-n classes to 10, we get 1171 and 459 videos for train and test sets respectively. However, the preprocessing step takes 40 minutes to complete. So, I decided to stick to the previously subsampled dataset. To make the readers fully aware of the low-data regime this example operates on, I have included an extensive note. 
* I tried with smaller architectures like MobileNetV3 but that did not result in much speed-up since most of the time is being spent on frame extraction (handled by OpenCV). 
* Upon trying smaller but better performing models like `EfficientNetB0`, I observed something weird. The performance was dropping consistently (+/- 10%). The data regime is too small to comment anything definitively but here are some thoughts:
  * We are primarily interested in extracting meaningful features. Networks like VGG16, InceptionV3 have a good history of being used in works that deal with image feature spaces. Some relevant works: [LPIPS](https://arxiv.org/abs/1801.03924), [Perceptual Loss](https://arxiv.org/abs/1603.08155), and the famous Deep Dream. 
  * Whereas EfficientNet family of models were found via neural architecture search constrained by aspects like minimal FLOP count, speedy training schedule, and low inference latency. I suspect this may have something to do with the disparity in the feature representations they produce. 

Here's the [full notebook](https://colab.research.google.com/github/sayakpaul/Action-Recognition-in-TensorFlow/blob/main/Video_Classification.ipynb).

fchollet commented:
Fantastic! Everything is looking good. Please add the generated files 👍

sayakpaul commented:
Thanks for all your help and feedback, @fchollet. I have added the files. The visualization GIF was probably embedded in the notebook but if you find anything off let me know.

fchollet reviewed (approved):
LGTM, thank you for the great contribution! I confirm the gif is being captured -- some guy shaving his beard.

fchollet commented:
Actually, there's a small issue -- it seems [these copyedits](https://github.com/keras-team/keras-io/pull/478/commits/cf71064b604420c58786a7dcf4b21462a3dddb1e) were dropped. Can you please create a new PR to add them back?
